### PR TITLE
feat: sync unit test impl

### PR DESCRIPTION
**Pull request by sitaram-kalluri**
**- What I did**
1. Write sync unit tests on the at_secondary_server.

**- How I did it**
1. Analyze the sync unit tests and write pre-conditions and assertions
2. Implement the unit tests in line with preconditions and assertions

**- How to verify it**
1. The sync unit tests should pass.
3. The unit tests should cover the sync functionality

**- Description for the changelog**
1. fix: Add unit tests for sync

TODO: There are 4 tests pending to implement. Out of which two tests need code refactoring and two need a discussion on the expected behaviour.

Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->

### CODE REVIEW

**Review by purnimavenkatasubbu — DISMISSED** [review]

**Review by srieteja — DISMISSED** [review]

File: packages/at_secondary_server/lib/src/verb/handler/sync_progressive_verb_handler.dart [hunk 1]
```diff
@@ -9,10 +9,14 @@ import 'package:at_secondary/src/verb/handler/abstract_verb_handler.dart';
 import 'package:at_secondary/src/verb/verb_enum.dart';
 import 'package:at_server_spec/at_server_spec.dart';
 import 'package:at_server_spec/at_verb_spec.dart';
+import 'package:meta/meta.dart';
 
 class SyncProgressiveVerbHandler extends AbstractVerbHandler {
   static SyncFrom syncFrom = SyncFrom();
 
+  @visibleForTesting
+  ByteBuffer syncBuffer = ByteBuffer(capacity: AtSecondaryConfig.syncBufferSize);
```
Comment (VJag) [inline packages/at_secondary_server/lib/src/verb/handler/sync_progressive_verb_handler.dart]: Can this be an issue when dealing with async and await behavior?

File: packages/at_secondary_server/test/sync_unit_test.dart [hunk 2]
```diff
@@ -0,0 +1,724 @@
...
+            atCommitLog!.getLatestCommitEntry('@alice:phone@alice');
+        expect(commitEntryList!.operation, CommitOp.UPDATE_META);
+        expect(commitEntryList.commitId, 1);
+      });
+
+      test(
+          'test to verify when the commit entry is delete, the key is deleted from hive keystore and entry added to commit log',
+          () async {
+        ///Precondition
+        /// 1. The serverCommitId is at 5
```
Comment (VJag) [inline packages/at_secondary_server/test/sync_unit_test.dart]: Comments and code are not in sync

File: packages/at_secondary_server/test/sync_unit_test.dart [hunk 3]
```diff
@@ -0,0 +1,724 @@
...
+        /// 2. The key already exists in cloud secondary
+        /// 3. delete commit entry is received
+        ///
+        /// Assertions:
+        /// 1. The should be deleted from the keystore
+        /// 2. The commit log should be updated
+        await secondaryPersistenceStore!
```
Comment (VJag) [inline packages/at_secondary_server/test/sync_unit_test.dart:182]: Where is the key "@alice:phone@alice" inserted before the delete? Should the commit ID and version be 1?

"Please check the code and comments and see that they are in sync"

File: packages/at_secondary_server/test/sync_unit_test.dart [hunk 4]
```diff
@@ -0,0 +1,724 @@
...
+        var atConnection = InboundConnectionImpl(null, inBoundSessionId);
+        atConnection.metaData.isAuthenticated = true;
+        var batchVerbParams = HashMap<String, String>();
+        batchVerbParams.putIfAbsent('json', () => batchRequestCommand);
+        await batchVerbHandler.processVerb(
+            response, batchVerbParams, atConnection);
+        List batchResponseList = jsonDecode(response.data!);
+
+        // Assertions
+        expect(batchResponseList.length, 4);
```
Comment (VJag) [inline packages/at_secondary_server/test/sync_unit_test.dart:261]: Great test (clap clap clap)

File: packages/at_secondary_server/test/sync_unit_test.dart [hunk 5]
```diff
@@ -0,0 +1,724 @@
...
+            response, batchVerbParams, atConnection);
+        List batchResponseList = jsonDecode(response.data!);
+        expect(batchResponseList.length, 3);
+        expect(batchResponseList[0]['response']['data'], '0');
+        expect(batchResponseList[1]['response']['error_code'], 'AT0003');
+        expect(batchResponseList[1]['response']['error_message'],
+            'Invalid syntax');
+        expect(batchResponseList[2]['response']['data'], '1');
+      });
+      test('Arch - test to verify when invalid json is sent in batch request',
```
Comment (VJag) [inline packages/at_secondary_server/test/sync_unit_test.dart]: Does this require an arch discussion? If yes, should we track it through a ticket? I thought we have the server throwing an internal server exception for this case. I mean we do handle invalid JSON already I recon.

File: packages/at_secondary_server/test/sync_unit_test.dart [hunk 6]
```diff
@@ -0,0 +1,724 @@
...
+        syncVerbParams.putIfAbsent(AT_FROM_COMMIT_SEQUENCE, () => '-1');
+        syncVerbParams.putIfAbsent('regex', () => 'buzz');
+        await syncProgressiveVerbHandler.processVerb(
+            response, syncVerbParams, atConnection);
+
+        List syncResponse = jsonDecode(response.data!);
+
+        expect(syncResponse.length, 2);
+        // As per design, all the public keys are not filtered when matching with regex.
+        expect(syncResponse[0]['atKey'], 'public:country.wavi@alice');
```
Comment (VJag) [inline packages/at_secondary_server/test/sync_unit_test.dart:434]: How is this key matching the regex 'buzz'?

File: packages/at_secondary_server/lib/src/verb/handler/sync_progressive_verb_handler.dart (hunk 1 above)
Comment (sitaram-kalluri) [reply packages/at_secondary_server/lib/src/verb/handler/sync_progressive_verb_handler.dart]: Resolved this comment. Moved the sync buffer into the processVerb method.

File: packages/at_secondary_server/test/sync_unit_test.dart [hunk 7]
```diff
@@ -0,0 +1,724 @@
...
+        var response = Response();
+        var inBoundSessionId = '_6665436c-29ff-481b-8dc6-129e89199718';
+        var atConnection = InboundConnectionImpl(null, inBoundSessionId);
+        atConnection.metaData.isAuthenticated = true;
+        var syncVerbParams = HashMap<String, String>();
+        syncVerbParams.putIfAbsent(AT_FROM_COMMIT_SEQUENCE, () => '-1');
+        await syncProgressiveVerbHandler.processVerb(
+            response, syncVerbParams, atConnection);
+        List syncResponse = jsonDecode(response.data!);
+        expect(syncResponse.length, 1);
```
Comment (VJag) [inline packages/at_secondary_server/test/sync_unit_test.dart:471]: Can we write this test in the following way:

1. We insert two keys in a way that only one of them fits into the buffer limit
     i.e size(key1) + size(key2) > buffer limit, however size(key1) < buffer limit
2.  Verify that the sync response contains only key1

Now increase the buffer limit in a way that accommodates both the keys
 i.e size(key1) + size(key2) < buffer limit
3.  Verify that the sync response contains both the keys

File: packages/at_secondary_server/test/sync_unit_test.dart (hunk 6 above)
Comment (sitaram-kalluri) [reply packages/at_secondary_server/test/sync_unit_test.dart:434]: When matching the regex, we look for keys that match the regex or if the key falls under the special key.

Below is the snippet
```dart
bool _acceptKey(String atKey, String regex) {
    return _isRegexMatches(atKey, regex) || _isSpecialKey(atKey);
  }
```

In this case, the key does not match the regex but matches the special key - because the key is public.

Below is the snippet for _specialKey method:
```dart
bool _isSpecialKey(String atKey) {
    return atKey.contains(AT_ENCRYPTION_SHARED_KEY) ||
        atKey.startsWith('public:') ||
        atKey.contains(AT_PKAM_SIGNATURE) ||
        atKey.contains(AT_SIGNING_PRIVATE_KEY);
  }
```

File: packages/at_secondary_server/test/sync_unit_test.dart [hunk 8]
```diff
@@ -0,0 +1,724 @@
...
+        /// 1. The key is created on server at time T1
+        /// 2. The key is expired at time T2 (T2 > T1); but key is still in keystore and not deleted
+        /// 3. The sync triggers at time T3 (T3 > T2)
+        ///
+        /// Things to discuss:
+        /// Though the key is expired (not deleted), the commit log do not have delete entry
+        /// So, the latest entry of the key is expired
+        /// And at the moment, we sync the expired key
+        ///
+        /// TODO: ** Needs refactoring:
```
Comment (VJag) [inline packages/at_secondary_server/test/sync_unit_test.dart]: @kalluriramkumar In one of the arch calls discuss all of the tickets raised / things we want to refactor for tests.

Comment (sitaram-kalluri) [reply packages/at_secondary_server/test/sync_unit_test.dart]: Sure @VJag

File: packages/at_secondary_server/test/sync_unit_test.dart (hunk 7 above)
Comment (sitaram-kalluri) [reply packages/at_secondary_server/test/sync_unit_test.dart:471]: This change is implemented.

File: packages/at_secondary_server/test/sync_unit_test.dart (hunk 3 above)
Comment (sitaram-kalluri) [reply packages/at_secondary_server/test/sync_unit_test.dart:182]: This change is implemented

File: packages/at_secondary_server/test/sync_unit_test.dart (hunk 2 above)
Comment (sitaram-kalluri) [reply packages/at_secondary_server/test/sync_unit_test.dart]: This change is implemented

**Comment by gkc** [comment]
Is this still draft, or is it ready for review?

File: packages/at_secondary_server/test/sync_unit_test.dart (hunk 6 above)
Comment (VJag) [reply packages/at_secondary_server/test/sync_unit_test.dart:434]: That logic seems incorrect. But we will deal with that later.

**Review by VJag — DISMISSED** [review]

**Review by gkc — CHANGES_REQUESTED** [review]
@sitaram-kalluri As per discussion the other day, I was happy for this to be merged after the review we had once all the TODOs have been added etc. 

Sadly, you just did a force-push which means I need to re-review the whole PR, or can you summarize for me what the actual changes were please? 
image

**Comment by sitaram-kalluri** [comment]
> @sitaram-kalluri As per discussion the other day, I was happy for this to be merged after the review we had once all the TODOs have been added etc.
> 
> Sadly, you just did a force-push which means I need to re-review the whole PR, or can you summarize for me what the actual changes were please? image

@gkc: The commit is to pull the latest merges of the trunk into this branch. Going forward I will avoid force push to the branch.

**Review by gkc — APPROVED** [review]
